README.md
# rusty-battery

[![crates.io](https://img.shields.io/crates/v/rusty-battery?logo=rust)](https://crates.io/crates/rusty-battery)
[![pages-build-deployment](https://github.com/kucera-lukas/rusty-battery/actions/workflows/pages/pages-build-deployment/badge.svg)](https://rustybattery.lukaskucera.com)
[![Continuous Integration](https://github.com/kucera-lukas/rusty-battery/actions/workflows/ci.yml/badge.svg)](https://github.com/kucera-lukas/rusty-battery/actions/workflows/ci.yml)

CLI tool which notifies you when laptop battery reaches a threshold.

# Why should you use `rusty-battery`?

If you aren't able to set start/stop charge threshold
(for example via [TLP](https://linrunner.de/tlp/)) but would still like to
make sure that your battery won't exceed your preferred threshold.
`rusty-battery` can let you know when your battery reached the threshold by
showing a desktop notification and optionally pinging your
[KDE Connect](https://kdeconnect.kde.org/) devices.

# Features

- [notify](#notify)
- [batteries](#batteries)
- [kde-connect-devices](#kde-connect-devices)

## notify

Notify whenever battery percentage exceeds the given threshold

USAGE:

    rusty-battery notify [OPTIONS]

OPTIONS:

    --disable-desktop
            Disable desktop notifications

            Specify this flag if you don't want desktop notifications to be shown whenever the
            chosen battery percentage exceeds the given threshold.

    -h, --help
            Print help information

    --kde-connect [<KDE_CONNECT_NAMES>...]
            KDE Connect device names

            If this value is not present, KDE Connect will not be used.

            If this value is empty, all of the KDE Connect devices will be pinged.

    -m, --model <MODEL>
            Battery model name

            If this value is omitted and only battery device is found for the current device, that
            one will be used.

            Otherwise, please use the `batteries` subcommand to get a list of all battery devices to
            get the model of the wanted battery device which should be monitored.

    -q, --quiet
            Less output per occurrence

    --refresh-secs <REFRESH_SECS>
            Number of seconds to wait before refreshing battery device data

            After every battery device refresh, its data will be checked. Notifications will be sent
            everytime they should be, based on the new refreshed battery device data.

            [default: 30]

    -t, --threshold <THRESHOLD>
            Battery charge threshold

            Whenever the chosen battery device reaches this charge threshold and will be charging,
            notifications will be sent, alerting that the charger should be unplugged.

            [minimum: 0] [maximum: 100]

            [default: 80]

    -v, --verbose
            More output per occurrence

    -V, --version
            Print version information

## batteries

List all available batteries of the current device

USAGE:

    rusty-battery batteries [OPTIONS]

OPTIONS:

    -h, --help       Print help information
    -q, --quiet      Less output per occurrence
    -v, --verbose    More output per occurrence
    -V, --version    Print version information

## kde-connect-devices

List all available KDE Connect devices

USAGE:

    rusty-battery kde-connect-devices [OPTIONS]

OPTIONS:

    -h, --help       Print help information
    -q, --quiet      Less output per occurrence
    -v, --verbose    More output per occurrence
    -V, --version    Print version information

# Installation

## From [crates.io](https://crates.io/crates/rusty-battery)

```sh
cargo install rusty-battery
```

## From [source](https://github.com/kucera-lukas/rusty-battery)

1. Clone the repository

```sh
git clone git@github.com:kucera-lukas/rusty-battery.git
```

2. Change directory

```sh
cd rusty-battery
```

3. Install with cargo

```sh
cargo install --path .
```

## From [release page](https://github.com/kucera-lukas/rusty-battery/releases)

Download a binary of the
[latest release](https://github.com/kucera-lukas/rusty-battery/releases/latest)
and move it to a directory which is in your `$PATH`.
You may need to change the binary's permissions by running:

```sh
chmod +x rusty-battery
```

If there are any problems with the pre-compiled binaries, file an issue.

# Usage tips

This tool is best used when set up as a background task.

### Setup with `cron`

1. Open crontab

```sh
crontab -e
```

2. Paste in `@reboot rusty-battery notify [YOUR OPTIONS]`
3. Save and exit the text editor, you should
   see `crontab: installing new crontab` in your terminal
4. Reboot the system

```sh
reboot
```

### Logging

1. Choose the log verbosity via the `-v` or `--verbose` flag
2. Append it to the `rusty-battery` command
3. Redirect output via `>> /path/to/log/file 2>&1`
4. Check all logs via

```sh
more /path/to/log/file
```

- Check live logs:

```sh
tail -f /path/to/log/file
```

- `2>&1` [explanation](https://stackoverflow.com/questions/818255/in-the-shell-what-does-21-mean)

### Debugging

- [Here is a useful thread](https://askubuntu.com/questions/23009/why-crontab-scripts-are-not-working)
  for crontab debugging
- To check that `rusty-battery` is running you can use

```sh
ps aux | grep -e rusty-battery
```

- To kill the job you can use (`$PID` can be found via the previous command):

```sh
kill $PID
```

# Device support

Tested on:

- OS: Fedora 34, 35, 36
- DE: Plasma


CHANGELOG.md
# Changelog

## [0.7.0](https://github.com/kucera-lukas/rusty-battery/compare/v0.6.1...v0.7.0) (2023-05-27)

### Features

- systemd configuration ([#195](https://github.com/kucera-lukas/rusty-battery/issues/195)) ([8268a8b](https://github.com/kucera-lukas/rusty-battery/commit/8268a8b81138b92ed531a26263e2dcd7d711cbbe))

## [0.6.1](https://github.com/kucera-lukas/rusty-battery/compare/v0.6.0...v0.6.1) (2022-10-02)

### Miscellaneous Chores

- update readme file path in Cargo.toml ([932dc81](https://github.com/kucera-lukas/rusty-battery/commit/932dc8140e29eeda76ec04eed8f7ff95980d232f))

## [0.6.0](https://github.com/kucera-lukas/rusty-battery/compare/v0.5.0...v0.6.0) (2022-10-02)

### Features

- add custom notification summary and body ([#138](https://github.com/kucera-lukas/rusty-battery/issues/138)) ([3012e90](https://github.com/kucera-lukas/rusty-battery/commit/3012e90568e8c7a4e4669eaeca0f8f9390c1c529))
- add Google Analytics ([#139](https://github.com/kucera-lukas/rusty-battery/issues/139)) ([5f3ff72](https://github.com/kucera-lukas/rusty-battery/commit/5f3ff72b030f1c17abeab8ec3fdbd3373a8b0e30))
- refactor notification module ([#136](https://github.com/kucera-lukas/rusty-battery/issues/136)) ([1c3dead](https://github.com/kucera-lukas/rusty-battery/commit/1c3dead17e409ce388803554d9bb6e00e58fa680))

## [0.5.0](https://github.com/kucera-lukas/rusty-battery/compare/v0.4.0...v0.5.0) (2022-09-30)

### Features

- better input validation ([#132](https://github.com/kucera-lukas/rusty-battery/issues/132)) ([8ba5794](https://github.com/kucera-lukas/rusty-battery/commit/8ba57946ff375c14369352686b41b90e31b5ad85))
- migrate to clap v4 ([#133](https://github.com/kucera-lukas/rusty-battery/issues/133)) ([48cc540](https://github.com/kucera-lukas/rusty-battery/commit/48cc540086a5ca947b69757cd96c8085ec5528ec))
- remove notification on SIGINT, SIGTERM and SIGHUP ([#130](https://github.com/kucera-lukas/rusty-battery/issues/130)) ([64a4f66](https://github.com/kucera-lukas/rusty-battery/commit/64a4f66188aa80a3e2e3d51828e858ed842fc7e3))
- warn about missing kde connect devices on startup ([#128](https://github.com/kucera-lukas/rusty-battery/issues/128)) ([2ba1c0f](https://github.com/kucera-lukas/rusty-battery/commit/2ba1c0f1827c12700ed4ff356323ce5f9c75763f))

## [0.4.0](https://github.com/kucera-lukas/rusty-battery/compare/v0.3.0...v0.4.0) (2022-09-27)

### Features

- add cli flag to disable desktop notifications ([#124](https://github.com/kucera-lukas/rusty-battery/issues/124)) ([ca2060f](https://github.com/kucera-lukas/rusty-battery/commit/ca2060f566e15ca69442e2eb52bed8ec95fe7267))
- add prefix to logging messages ([#123](https://github.com/kucera-lukas/rusty-battery/issues/123)) ([d89f690](https://github.com/kucera-lukas/rusty-battery/commit/d89f69016ac9f0f09796177fd22f42d26ad90663))
- log information connected to cli startup ([#125](https://github.com/kucera-lukas/rusty-battery/issues/125)) ([3881a6e](https://github.com/kucera-lukas/rusty-battery/commit/3881a6eb06a66bf5eeba13fc63ca43b0844e748b))
- migrate to clap v3 ([#121](https://github.com/kucera-lukas/rusty-battery/issues/121)) ([96f9028](https://github.com/kucera-lukas/rusty-battery/commit/96f90287214a3c39ce2b1c6445bb34c15abfeff2))

## [0.3.0](https://www.github.com/kucera-lukas/rusty-battery/compare/v0.2.1...v0.3.0) (2022-09-25)

### Features

- close notification automatically ([#116](https://www.github.com/kucera-lukas/rusty-battery/issues/116)) ([0b0a1fc](https://www.github.com/kucera-lukas/rusty-battery/commit/0b0a1fc3dc1e2291721272e6e41494d6d69e0252))
- update continuous integration ([#117](https://www.github.com/kucera-lukas/rusty-battery/issues/117)) ([2675d37](https://www.github.com/kucera-lukas/rusty-battery/commit/2675d373136450bb148760b2043f34b80cac2f89))
- update docs ([#120](https://www.github.com/kucera-lukas/rusty-battery/issues/120)) ([be07421](https://www.github.com/kucera-lukas/rusty-battery/commit/be07421dea49a422621ea461411f7fbd22fbaf40))

### [0.2.1](https://www.github.com/kucera-lukas/rusty-battery/compare/v0.2.0...v0.2.1) (2021-10-28)

### Miscellaneous Chores

- release 0.2.1 ([23f59b9](https://www.github.com/kucera-lukas/rusty-battery/commit/23f59b99527c695855db25419e9b54758e5e4889))

## [0.2.0](https://www.github.com/kucera-lukas/rusty-battery/compare/v0.1.3...v0.2.0) (2021-10-28)

### Miscellaneous Chores

- release 0.2.0 ([0861c7a](https://www.github.com/kucera-lukas/rusty-battery/commit/0861c7a082780bda8a33f47830830699d2185ac7))


src/cli.rs
use clap::{Parser, Subcommand};

use crate::parser;

/// Tool to help you care about your device's battery health.
#[derive(Parser, Debug)]
#[command(author, version, about)]
#[command(propagate_version = true)]
pub struct Cli {
    /// Control log level with `--verbose` and `--quiet` flags.
    #[clap(flatten)]
    pub verbose: clap_verbosity_flag::Verbosity,

    #[command(subcommand)]
    pub cmd: Command,
}

#[derive(Subcommand, Debug, PartialEq, Eq)]
pub enum Command {
    /// Notify whenever battery percentage exceeds the given threshold.
    Notify {
        /// Battery charge threshold
        ///
        /// Whenever the chosen battery device reaches this charge threshold
        /// and will be charging, notifications will be sent, alerting that
        /// the charger should be unplugged.
        ///
        /// [minimum: 0] [maximum: 100]
        #[arg(
        short,
        long,
        value_parser = parser::threshold,
        default_value_t = 80
        )]
        threshold: u8,

        /// Battery model name
        ///
        /// If this value is omitted and only battery device is found
        /// for the current device, that one will be used.
        ///
        /// Otherwise, please use the `batteries` subcommand
        /// to get a list of all battery devices to get the model of the
        /// wanted battery device which should be monitored.
        #[arg(short, long)]
        model: Option<String>,

        /// Number of seconds to wait before refreshing battery device data
        ///
        /// After every battery device refresh, its data will be checked.
        /// Notifications will be sent everytime they should be, based on the
        /// new refreshed battery device data.
        #[arg(long, default_value_t = 30)]
        refresh_secs: u64,

        /// Notification summary
        ///
        /// Supported variables: THRESHOLD, CHARGE_STATE, MODEL, REFRESH_SECS
        ///
        /// Reference these variables in your summary like shell environment
        /// variables with the '$' prefix.
        #[arg(long, default_value_t = String::from("Charge limit warning"))]
        summary: String,

        /// Notification body
        ///
        /// Supported variables: THRESHOLD, CHARGE_STATE, MODEL, REFRESH_SECS
        ///
        /// Reference these variables in your body like shell environment
        /// variables with the '$' prefix.
        #[arg(
            long,
            default_value_t = String::from("Battery percentage reached the \
            $THRESHOLD% threshold, please unplug your charger")
        )]
        body: String,

        /// KDE Connect device names
        ///
        /// If this value is not present, KDE Connect will not be used.
        ///
        /// If this value is empty,
        /// all of the KDE Connect devices will be pinged.
        #[arg(long = "kde-connect", num_args = 0..=255)]
        kde_connect_names: Option<Vec<String>>,

        /// Disable desktop notifications
        ///
        /// Specify this flag if you don't want desktop notifications
        /// to be shown whenever the chosen battery percentage exceeds the
        /// given threshold.
        #[arg(long)]
        disable_desktop: bool,
    },
    /// List all available batteries of the current device.
    Batteries,
    /// List all available KDE Connect devices.
    KDEConnectDevices,
}

pub fn parse() -> Cli {
    Cli::parse()
}

#[cfg(test)]
mod tests {
    use super::Cli;

    #[test]
    fn test_cli() {
        use clap::CommandFactory;

        Cli::command().debug_assert();
    }
}
